packages/vhd-lib/src/createVhdDirectoryFromStream.js

```diff
@@ -0,0 +1,178 @@
+    buffer = buffer.slice(buffer.length - maxSize)
+  }
+  return buffer
+}
+
+async function* parse(stream) {
+  let bytesRead = 0
+
+  // handle empty space between elements
+  // ensure we read stream in order
+  async function read(offset, size) {
+    assert(bytesRead <= offset, `offset is ${offset} but we already read ${bytesRead} bytes`)
+    if (bytesRead < offset) {
+      // empty spaces
+      await read(bytesRead, offset - bytesRead)
+    }
+    const buf = await readChunk(stream, size)
+    assert.strictEqual(buf.length, size, `read ${buf.length} instead of ${size}`)
+    bytesRead += size
+    return buf
+  }
+
+  const bufFooter = await read(0, FOOTER_SIZE)
+
+  const footer = buildFooter(bufFooter)
+  yield { type: 'footer', footer, offset: 0 }
+
+  const bufHeader = await read(FOOTER_SIZE, HEADER_SIZE)
+  const header = buildHeader(bufHeader, footer)
+
+  yield { type: 'header', header, offset: SECTOR_SIZE }
+  const blockSize = header.blockSize
+  assert.strictEqual(blockSize % SECTOR_SIZE, 0)
+
+  const blockBitmapSize = computeBlockBitmapSize(blockSize)
+  const blockAndBitmapSize = blockBitmapSize + blockSize
+
+  const index = []
+
+  for (const parentLocatorId in header.parentLocatorEntry) {
+    const parentLocatorEntry = header.parentLocatorEntry[parentLocatorId]
+    // empty parent locator entry, does not exist in the content
+    if (parentLocatorEntry.platformDataSpace === 0) {
+      continue
+    }
+    index.push({
+      ...parentLocatorEntry,
+      type: 'parentLocator',
+      offset: parentLocatorEntry.platformDataOffset,
+      size: parentLocatorEntry.platformDataLength,
+      id: parentLocatorId,
+    })
+  }
+
+  const batOffset = header.tableOffset
+  const batSize = Math.max(1, Math.ceil((header.maxTableEntries * 4) / SECTOR_SIZE)) * SECTOR_SIZE
+
+  index.push({
+    type: 'bat',
+    offset: batOffset,
+    size: batSize,
+  })
+
+  // sometimes some parent locator are before the BAT
+  index.sort((a, b) => a.offset - b.offset)
+  while (index.length > 0) {
+    const item = index.shift()
+    const buffer = await read(item.offset, item.size)
+    if (item.type === 'bat') {
+      // found the BAT : read it and ad block to index
+      for (let blockCounter = 0; blockCounter < header.maxTableEntries; blockCounter++) {
+        const batEntrySector = buffer.readUInt32BE(blockCounter * 4)
+        // unallocated block, no need to export it
+        if (batEntrySector !== BLOCK_UNUSED) {
+          const batEntryBytes = batEntrySector * SECTOR_SIZE
+          // ensure the block is not before the bat
+          assert.ok(batEntryBytes >= batOffset + batSize)
+          index.push({
+            type: 'block',
+            id: blockCounter,
+            offset: batEntryBytes,
+            size: blockAndBitmapSize,
+          })
+        }
+      }
+      // sort again index to ensure block and parent locator are in the right order
+      index.sort((a, b) => a.offset - b.offset)
```

Comment:
As this will be done an huge number of times, I think it makes sense to optimize it with `lodash.sortedIndex` and `Array#splice`.

Reply:
My bad, you only call it twice, it's good enough for me.